src/CommonLib/Ntlm/LdapNative.cs

```diff
@@ -0,0 +1,346 @@
+﻿#nullable enable
+
+using Microsoft.Win32.SafeHandles;
+using SharpHoundCommonLib.Enums;
+using System;
+using System.Runtime.InteropServices;
+
+namespace SharpHoundCommonLib.Ntlm;
+
+public class SafeLdapHandle : SafeHandleZeroOrMinusOneIsInvalid
+{
+    internal SafeLdapHandle(IntPtr handle) : base(true)
+    {
+        this.handle = handle;
+    }
+
+    protected override bool ReleaseHandle()
+    {
+        if (!IsInvalid)
+        {
+            int result = NativeMethods.ldap_unbind_s(handle);
+            return result == 0;
+        }
+        return true;
+    }
+}
+
+public class LdapConnection : IDisposable
+{
+    private SafeLdapHandle? _handle;
+    private bool _disposed;
+
+    public SafeLdapHandle Handle
+    {
+        get
+        {
+            ThrowIfDisposed();
+            return _handle ?? throw new InvalidOperationException("LDAP connection not initialized");
+        }
+    }
+
+    private void ThrowIfDisposed()
+    {
+        if (_disposed)
+        {
+            throw new ObjectDisposedException(nameof(LdapConnection));
+        }
+    }
+
+    private string? GetServerError()
+    {
+        ThrowIfDisposed();
+
+        try
+        {
+            IntPtr errorMessagePtr;
+            int result = NativeMethods.ldap_get_option(
+                Handle,
+                LdapOption.ServerError,
+                out errorMessagePtr
+            );
+
+            if (result != 0 || errorMessagePtr == IntPtr.Zero)
+                return null;
+
+            try
+            {
+                return Marshal.PtrToStringAnsi(errorMessagePtr);
+            }
+            finally
+            {
+                NativeMethods.ldap_memfree(errorMessagePtr);
+            }
+        }
+        catch
+        {
+            // If anything goes wrong getting the server error,
+            // return null rather than throw since we're already
+            // in error handling code
+            return null;
+        }
+    }
+
+    public void CheckError(int result, string operation)
+    {
+        if (result != 0)
```

Comment:
I think we could replace `0` with `LdapErrorCodes.Success`, but not sure if the return of the `ldap_set_option` system call is the same as the out value